### PR TITLE
First iteration of automated version check for e2e tests

### PR DESCRIPTION
**Pull request by holisticode**
Closes https://github.com/ava-labs/avalanche-cli/issues/420

### CODE REVIEW

File: tests/e2e/utils/versions_test.go [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+	}
+	return &testMapper{
+		app,
+		nil,
+		nil,
+		t,
+	}
+}
+
+func (m *testMapper) GetLatestAvagoByProtoVersion(app *application.Avalanche, rpcVersion int, url string) (string, error) {
```
Comment (dasconnor) [inline tests/e2e/utils/versions_test.go:50]: why are you passing in a URL?

Comment (dasconnor) [reply tests/e2e/utils/versions_test.go:50]: again, I'm not sure if you're comparing against `latest` so you might end up with some prerelease confusion

File: tests/e2e/utils/versions_test.go [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+		_, err = w.Write([]byte(m.currentContext.sourceAvago))
+	default:
+		m.t.Fatalf("Unexpected path URL for test server: %s\n", r.URL.Path)
+	}
+	if err != nil {
+		m.t.Fatal(err)
+	}
+}
+
+func TestGetVersionMapping(t *testing.T) {
```
Comment (dasconnor) [inline tests/e2e/utils/versions_test.go:112]: Sorry, this isn't going to be the most helpful review comment ever, but I am extremely confused as to what this file is testing. I would suggest labeling/naming each test case and moving some of the above logic into the test so that it looks a bit more linear. It's very hard to follow.

File: pkg/application/app.go [hunk 3]
```diff
@@ -127,6 +127,14 @@ func (app *Avalanche) GetKeyPath(keyName string) string {
 	return filepath.Join(app.baseDir, constants.KeyDir, keyName+constants.KeySuffix)
 }
 
+func (app *Avalanche) GetDownloader() Downloader {
+	return app.Downloader
+}
+
+func (app *Avalanche) GetAvalanchegoCompatibilityURL() string {
```
[the remark this answers is not in the document]
Comment (holisticode) [reply pkg/application/app.go:134]: Yes

File: pkg/vm/compatibility_test.go [hunk 4]
```diff
@@ -176,7 +177,8 @@ func TestGetLatestAvalancheGoByProtocolVersion(t *testing.T) {
 			app := application.New()
 			app.Downloader = mockDownloader
 
-			avagoVersion, err := GetLatestAvalancheGoByProtocolVersion(app, tt.rpc)
+			// TODO: this should be a different URL, a unit test should not need to access a live URL
```
[the remark this answers is not in the document]
Comment (holisticode) [reply pkg/vm/compatibility_test.go:180]: I think you're right, in some iteration I was mocking the app but not the Downloader, so it was hitting github. But this test mocks the Downloader.

File: tests/e2e/testcases/packageman/suite.go [hunk 5]
```diff
@@ -15,18 +15,20 @@ import (
...
-	soloSubnetEVMVersion1 = "v0.4.2"
-	soloSubnetEVMVersion2 = "v0.4.1"
-
-	soloAvagoVersion = "v1.9.1"
-
-	multipleAvagoSubnetEVM = "v0.4.3"
-	multipleAvagoVersion1  = "v1.9.2"
-	mulitpleAvagoVersion2  = "v1.9.3"
+var (
+	mapping map[string]string
```
[the remark this answers is not in the document]
Comment (holisticode) [reply tests/e2e/testcases/packageman/suite.go:21]: Is `binary2Version` ok?

File: tests/e2e/testcases/subnet/local/suite.go [hunk 6]
```diff
@@ -276,6 +287,7 @@ var _ = ginkgo.Describe("[Subnet Compatibility]", func() {
 	})
 
 	ginkgo.It("can't deploy conflicting vm versions", func() {
+		// TODO: These shouldn't be hardcoded either
```
[the remark this answers is not in the document]
Comment (holisticode) [reply tests/e2e/testcases/subnet/local/suite.go:290]: https://github.com/ava-labs/avalanche-cli/issues/488

File: tests/e2e/utils/versions.go [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+		}
+		rpcs = append(rpcs, kint)
+	}
+	sort.Sort(sort.Reverse(sort.IntSlice(rpcs)))
+
+	// iterate the rpc versions
+	// evaluate two avalanchego versions which are consecutive
+	// and run with the same RPC version.
+	// This is required for the for the "can deploy with multiple avalanchego versions" test
+	for _, v := range rpcs {
```
[the remark this answers is not in the document]
Comment (holisticode) [reply tests/e2e/utils/versions.go:141]: OK, done something.

File: tests/e2e/utils/versions_test.go (hunk 1 above)
Comment (holisticode) [reply tests/e2e/utils/versions_test.go:50]: Because the call for the production code (`vm.GetLatestAvalancheGoByProtocolVersion`) requires it.

File: tests/e2e/utils/versions_test.go (hunk 2 above)
Comment (holisticode) [reply tests/e2e/utils/versions_test.go:112]: I added detailed code documentation to the test.
The idea of the test is that it actually runs the `GetVersionMapping`. For that to be useful, it needs to download some data. The test fakes github by returning `compatibility.json` content from the test, and makes sure that the return from `GetVersionMapping` matches an expected set.

File: tests/e2e/utils/versions.go [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+	// get compatible versions for subnetEVM
+	// subnetEVMversions is a list of sorted EVM versions,
+	// subnetEVMmapping maps EVM versions to their RPC versions
+	subnetEVMversions, subnetEVMmapping, err := getVersions(mapper, models.SubnetEvm)
+	if err != nil {
+		return nil, err
+	}
+	// subnet-evm publishes its upcoming new version in the compatibility json
+	// before the new version is actually a downloadable release
+	subnetEVMversions = subnetEVMversions[1:]
```
[the remark this answers is not in the document]
Comment (holisticode) [reply tests/e2e/utils/versions.go:113]: OK. 

For avalanchego, the code will only download versions based on a RPC which we get from a VM - which means it will always download a version which has already been released.

And for the tests where avalanchego runs alone, we just download `latest`. So for this use case, I don't need to apply that thing to avalanchego.

File: tests/e2e/utils/versions_test.go [hunk 9]
```diff
@@ -336,13 +406,17 @@ func TestGetVersionMapping(t *testing.T) {
 	}
 
 	for i, tc := range testContexts {
+		// run the function, but use the testMapper,
```
Comment (dasconnor) [inline tests/e2e/utils/versions_test.go:409]: Can you add a `name` field to each test and call the `run` function so we can execute these as multiple tests? That we way we will know which one failed if one ever does. Like this:
image

**Review by dasconnor — COMMENTED** [review]
Very close! Just a couple small things

**Review by dasconnor — APPROVED** [review]
